src/librustc_resolve/lib.rs

```diff
@@ -2180,10 +2140,18 @@ impl<'a> Resolver<'a> {
                 // This is a path in the type namespace. Walk through scopes
                 // looking for it.
                 if let Some(def) = resolution {
-                    // Write the result into the def map.
-                    debug!("(resolving type) writing resolution for `{}` (id {}) = {:?}",
-                           path_names_to_string(path, 0), ty.id, def);
-                    self.record_def(ty.id, def);
+                    match def.base_def {
+                        Def::Mod(..) => {
```

Comment:
This particular `Def::Mod` is not a marker, but a real module encountered in type position.
As for `Def::Mod` used in fake resolutions for `<T>::A`, it's a status quo, I'd like to come up with some better solution, then used now, but not in this PR. Just using a separate marker for it doesn't make much difference, I did some quick experiments.